packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/macro_resolve.rs

```diff
@@ -0,0 +1,230 @@
+//! name or state variable name, the longest valid partial match is used and unmatched portions of a `path` are
+//! preserved for future use.
+
+use std::{collections::HashMap, iter, mem};
+
+use crate::dast::flat_dast::{FlatNode, UntaggedContent};
+
+use super::{
+    flat_dast::{FlatRoot, Index},
+    PathPart,
+};
+use serde::Serialize;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error, PartialEq)]
+pub enum ResolutionError {
+    #[error("No node identified by path")]
+    NoReferent,
+    #[error("Path referred to more than one node")]
+    NonUniqueReferent,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct RefResolution {
+    pub node_idx: Index,
+    pub unresolved_path: Option<Vec<PathPart>>,
+}
+
+/// Status of a pointer referring to children of an element.
+#[derive(Debug)]
+enum Ref {
+    Unique(Index),
+    Ambiguous(Vec<Index>),
+}
+
+/// A `Resolver` is used to lookup elements by path/name. It constructs a search index
+/// upon construction. If the underlying `FlatRoot` changes, a new `Resolver` should be
+/// recreated.
+#[derive(Debug)]
+pub struct Resolver {
+    /// List of the parent of a node at a given index.
+    node_parent: Vec<Option<Index>>,
+    /// Map of all the names that are accessible (as descendants) from a given node.
+    name_map: Vec<HashMap<String, Ref>>,
+}
+
+impl Resolver {
+    pub fn from_flat_root(flat_root: &FlatRoot) -> Self {
+        Resolver {
+            node_parent: flat_root.nodes.iter().map(|node| node.parent()).collect(),
+            name_map: Self::build_name_map(flat_root),
+        }
+    }
+
+    /// Search for a node specified by `path` starting from `origin`. This algorithm searches first
+    /// for the nearest parent that has an element with `path[0].name` as its name. Then it tries
+    /// to match as much of `path[1..]` as possible. A match is returned along with any unmatched
+    /// path parts.
+    ///
+    /// For example, consider
+    /// ```xml
+    /// <a name="x">
+    ///     <b name="y">
+    ///         <c name="z" />
+    ///     </b>
+    /// </a>
+    /// <d name="y" />
+    /// ```
+    /// - Matching `x` from anywhere returns the index of `<a />`.
+    /// - Matching `y` from `<c />` returns the index of `<b />`.
+    /// - Matching `y` from `<d />` returns an error because there are multiple `y`s accessible from that position.
+    /// - Matching `y.z.w` from `<b />` returns the index of `<c />` along with `.w` as the unresolved path.
+    ///
+    /// If an index appears in the path (e.g. `a[2].b`), the search stops and the remaining path is returned.
+    /// If there was a partial match of the indexed item, the unresolved path will list `name` as an empty string.
+    /// E.g., matching `y.w[2]` from `<b />` returns the index of `<d />` along with `.w[2]` as the unresolved path
+    /// and matching `y[2]` from `<b />` returns the index of `<d />` along with `.[2]` as the unresolved path.
+    pub fn resolve<T: AsRef<[PathPart]>>(
```

Comment:
What are we gaining by making the argument `AsRef<[PathPart]>` rather than `&[PathPart]`? I read some of the documentation `AsRef`, but couldn't figure out the benefit. It looks like it is current called on `&Vec<PathPath>`, for which the reference to the slice would work.

Reply:
`AsRef` allows passing anything that can be converted into a reference, so it allows both `Vec` and `&Vec`

Reply:
OK, like the example with a function taking both`String` and `&str`.